### PR TITLE
dashboard-KIT-data-quality

### PR DESCRIPTION
**Pull request by EmnaTay**

### CODE REVIEW

**Comment by EmnaTay** [comment]
hello! to understand better how to handle the files can you tell me how to process the .con files or tell me if you processed them in a specific file that i can look at? Thanks in advance.

**Comment by hzaatiti-NYU** [comment]
Let me know if you need additional information

**Comment by EmnaTay** [comment]
okay noted, thank you !

**Comment by hzaatiti-NYU** [comment]
Next short task:
- fix reading the .con in generate_snr_dashboard
- (optional) fix the box_script in another branch and make a new pull request (just add load_dot_env from a file)
- use script https://github.com/hzaatiti-NYU/meg-pipeline/blob/e20e51718acbe0c7616f2cdf97739ec5ebdff353/pipeline/plot_data.py to plot the raw_data structure of the empty_room.con (no remote code required, only locally)
- download https://mne.tools/stable/install/installers.html#installers mne standalone installer
- run mnekit2fiff (use tutorial from doc page)
- run plot on .fif file to get a feel of the data
- make the script to compute the average for each SQUID channel (in MNE labeled from MEG001 to MEG208, and excluding MEG092 (which is a broken sensor)

**Comment by EmnaTay** [comment]
Hello! I just finished all the tasks (except the optional one). If you would  like to add  any other tasks let me know, I would spend the rest of the day on the optional. 
Thanks in advance!

**Comment by Hzaatiti** [comment]
Hi Emna, please resync your branch to get the latest commits

File: docs/source/dashboards/generate_snr_dashboard.py [hunk 1]
```diff
@@ -87,28 +77,56 @@ def create_snr_plot(snr_values, output_file):
...
 
     # Path to your .fif file
-    #MEG_DATA = display_env_variable('MEG_DATA')
-    #file_path = '\empty-room\sub-emptyroom\meg-kit\empty-room-test_28_June_2024-raw_NO_OPM-raw.fif'
-    file_path = ('dashboards/data/empty-test.con')
+    # MEG_DATA = display_env_variable('MEG_DATA')
+    # file_path = '\empty-room\sub-emptyroom\meg-kit\empty-room-test_28_June_2024-raw_NO_OPM-raw.fif'
+    file_path = "C:/Users\Admin/meg-pipeline/docs/source/dashboards/data/kit_data/empty-test.con"
+
+    file_path_fif = "C:/Users\Admin/meg-pipeline/docs/source/dashboards/data/kit_data/empty-test.fif"
```
Comment (Hzaatiti) [inline docs/source/dashboards/generate_snr_dashboard.py:87]: Ditto

File: docs/source/dashboards/generate_snr_dashboard.py [hunk 2]
```diff
@@ -87,28 +77,56 @@ def create_snr_plot(snr_values, output_file):
...
+    ###channels_to_exclude = ["MEG092"]
+    raw.pick_types(
+        meg=True, eeg=False
+    )  # Pick only MEG channels and exclude EEG channels
+    raw.info["bads"]
+    data, times = raw[:]
+    # Compute the average of each channel
+    # The shape of data will be (n_channels, n_times)
+    averages = np.mean(data, axis=1)
+
```
Comment (Hzaatiti) [inline docs/source/dashboards/generate_snr_dashboard.py:103]: please correct comments grammar

File: docs/source/dashboards/generate_snr_dashboard.py [hunk 3]
```diff
@@ -87,28 +77,56 @@ def create_snr_plot(snr_values, output_file):
...
+    # raw_data = load_fif_data(file_path)
     raw_data = remove_zero_channels(raw_data)
 
-    with open('dashboards/data/empty_room_data.pkl', 'wb') as file:
+    with open("dashboards/data/empty_room_data.pkl", "wb") as file:
         pickle.dump(raw_data, file)
 
-    with open('dashboards/data/empty_room_data.pkl', 'rb') as file:
+    with open("dashboards/data/empty_room_data.pkl", "rb") as file:
         raw_data = pickle.load(file)
```
Comment (Hzaatiti) [inline docs/source/dashboards/generate_snr_dashboard.py:126]: i don't think we need to save a .pkl anymore is that right?
having the .con on box should be enough to compute the SNR from it directly

**Review by Hzaatiti — CHANGES_REQUESTED** [review]

**Comment by Hzaatiti** [comment]
Hi Emna, please change for "Ready to Review" after addressing my comments above, thank you

**Comment by EmnaTay** [comment]
Okay will do, thanks!

**Comment by EmnaTay** [comment]
Hello Dr. Hadi, I think I corrected the mistakes you pointed out yesterday but I am still problem is with the box-script I added two lines that load the .env file into the script directly but I am still  getting decoding errors at line 18, Would like your help with that! Thanks in advance!

**Comment by EmnaTay** [comment]
Exception has occurred: ValueError
Could not deserialize key data. The data may be in an incorrect format or it may be encrypted with an unsupported algorithm.
  File "C:\Users\Admin\Downloads\meg-pipeline\docs\source\dashboards\box_script.py", line 17, in <module>
    auth = JWTAuth(
ValueError: Could not deserialize key data. The data may be in an incorrect format or it may be encrypted with an unsupported algorithm.

**Comment by Hzaatiti** [comment]
tasks:
- formulas (define 2 in time-domain, average, variance) ==> Changing the   data_quality_dashboards.rst
- load a .con, compute the 2 formulas, display them with the time of the file (let us suppose that the name of the.con has a name)
- have the SNR's displayed on a table, one graph with time on x-axis, and the SNR value for a type on y- (different colors for different SNR metric)

(optional) repeat it for all .cons and have it plotted,

**Comment by EmnaTay** [comment]
Ready for review:

-Box-script.py: added a function that gets the .con files from all the folders in the Box database and has them located in the '/data'

-processing-con-files-for-table: gets all the data (.con files) and creates a csv file that contains {file_name, average over all channels, variance over all channels, date}, (Note: date is extracted from the file name, but not all files have a date; you will find that some dates are marked as unknown). The results are saved as a.csv that gets displayed later, and the results are saved as a simple line plot that is also displayed and added to the same 'data-quality-dashboard.rst'.

-under scripts added a script that converts all.csv files in (dashboards/data) to .rst files. The conversion was made to be able to add math symbols. 

-You will find that I added a function to automate the running of the script described above, but it kept raising errors for the math symbols, so I commented it out in line 113.

Good day!

**Comment by Hzaatiti** [comment]
Hi Emna, amazing work, this is looking great!

- I tested locally box_script, works well 
- tested processing_con_files_for_tables generation works well, it supposed 

Next tasks:

- merge 9-dashboard and dashboards, have the code in a separate directory than the .rst, for example called "dashboard_generation_scripts"
- address remarks on the scripts
- check if there's a way to retrieve the "Date modified" field, which usually, indicates the time this data has been generated by the KIT system, use this field for the time from now on instead of file name  (Note: if this is too tricky to do, please skip this task)
-

**Pull request merged**